### PR TITLE
CEP-001: Definition of the proposal process

### PR DESCRIPTION
**Pull request by Tobychev**
This is a first draft of the "root"  CEP that defines the purpose, scope and process for further proposals. Basically I took the [gammapy](https://docs.gammapy.org/dev/development/pigs/pig-001.html) and [Astropy](https://github.com/astropy/astropy-APEs/blob/main/APE1.rst) documents outlining their change proposal process and mashed them together, trying to abbreviate and restructure as seemed reasonable. 

This is very much a draft, both in terms of the actual process proposed and because I don't know the markup magic so some things that are supposed to be links are broken. 

The hope is that it will nevertheless sever as a useful basis for discussion.

### CODE REVIEW

File: docs/development/ceps/cep-001.rst [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+* Discussion:  NA
+* Date accepted: NA
+* Last revised: Febuary 10, 2023
+* Author: Tomas Bylund
+* Created: Febuary 10, 2023
+
+Abstract
+========
+
+CEP stands for "Ctapipe Enhancment Proposal". This is CEP 1, which
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:18]:
```suggestion
CEP stands for "ctapipe Enhancement Proposal". This is CEP 1, which
```

File: docs/development/ceps/cep-001.rst [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+describes the purpose of writing CEPs as well as giving some guidelines
+on how CEPs are authored, discussed and reviewed.
+
+An CEP is a design document providing information to the community, or
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:22]:
```suggestion
A CEP is a design document providing information to the community, or 
```

File: docs/development/ceps/cep-001.rst [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+
+Abstract
+========
+
+CEP stands for "Ctapipe Enhancment Proposal". This is CEP 1, which
+describes the purpose of writing CEPs as well as giving some guidelines
+on how CEPs are authored, discussed and reviewed.
+
+An CEP is a design document providing information to the community, or 
+describing a new feature for ctapipe or its processes or environment.
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:23]: only new features? I'd also include larger changes or design decisions

File: docs/development/ceps/cep-001.rst [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+on how CEPs are authored, discussed and reviewed.
+
+An CEP is a design document providing information to the community, or 
+describing a new feature for ctapipe or its processes or environment. 
+The CEP should provide a concise technical specification of the feature
+and a rationale for the feature.
+
+We intend CEPs to be the primary mechanisms for proposing major new features,
+for collecting community input on an issue, and for documenting the design
+decisions that have gone into Python. The CEP author is responsible for
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:29]:
```suggestion
decisions that have gone into ctapipe. The CEP author is responsible for
```

File: docs/development/ceps/cep-001.rst [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+
+Because the CEPs are maintained as text files in a versioned repository
+(indirectly since this wiki is versioned within github), their revision
+history is the historical record of the feature proposal
+
+
+What is a CEP?
+==============
+
+CTApipe Enhancment Proposals are short documents proposing a major
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:40]:
```suggestion
ctapipe Enhancment Proposals are short documents proposing a major
```

Comment (maxnoe) [reply docs/development/ceps/cep-001.rst:40]: I like this sentence much better than the above introduction sentence that only mentions new features. This is a bit duplicated, so maybe just use the one here?

File: docs/development/ceps/cep-001.rst [hunk 6]
```diff
@@ -0,0 +1,243 @@
...
+implement all of it.  If you're not sure if you should write a CEP, please don't! 
+Instead bring the topic up in discussions first with other ctapipe developers, or on
+the mailing list, to get some initial feedback. This will let you figure out if writing 
+a CEP will be helpful or not to realise your proposal.
+
+Writing a CEP
+=============
+
+The CEP process begins with a new idea for ctapipe. It is highly recommended
+that a single CEP contain a single key proposal or new idea. Note that small
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:75]:
```suggestion
that a single CEP contains a single key proposal or new idea. Note that small
```

File: docs/development/ceps/cep-001.rst [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+enhancements or patches as a rule don't need a CEP and can be injected into the
+ctapipe development workflow with a patch submission to the ctapipe issue
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:77]:
```suggestion
ctapipe development workflow via the ctapipe issue
```

File: docs/development/ceps/cep-001.rst [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+tracker. The more focused the CEP, the more successful it tends to be. If in
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:78]:
```suggestion
tracker or opening a pull request. The more focused the CEP, the more successful it tends to be. If in
```

File: docs/development/ceps/cep-001.rst [hunk 9]
```diff
@@ -0,0 +1,243 @@
...
+Once a complete draft for a CEP is written, it can be submitted for comunity discussion 
+by opening a pull request adding a RST file to the  ``docs/development/ceps`` folder 
+in the main ctapipe repository.
+
+Each CEP must have a advocate -- someone who beyond writing the CEP using the style and 
+format described above, shepherds the discussions in the appropriate forums, and attempts
+to build community consensus around the idea. This sheparding function is usually done by
+the author, but the role can be filled by any interested party.  Most discussions concerning
+ctapipe will happen by talking to each other directly (calls or face-to-face), or online on 
+the mailing list or Github.
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:158]: remove mailing list

File: docs/development/ceps/cep-001.rst [hunk 10]
```diff
@@ -0,0 +1,243 @@
...
+Anyone is welcome to review it and is encouraged to share their thoughts
+in the discussion!
+
+Please note that Github hides inline comments after they have been edited,
+so we suggest that you use inline comments for minor points like spelling 
+mistakes only. Put your main feedback as normal comments in the "Conversation"
+tab, so that for someone reading the discussion later they will see your 
+comment directly.
+
+The final decision on any CEP is made by the ctapipe coordination committee.
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:177]: That doesn't exist yet, so for now I'd write "core developer team"

File: docs/development/ceps/cep-001.rst [hunk 11]
```diff
@@ -0,0 +1,243 @@
...
+We expect that in most cases, the people participating in the CEP review will
+reach a consensus and the coordination committee will follow the outcome
+of the public discussion. But in unusual cases where disagreement remains,
+the coordination committee will talk to the people involved in the discussion
```
Comment (maxnoe) [inline docs/development/ceps/cep-001.rst:181]: see above

File: docs/development/ceps/cep-001.rst (hunk 3 above)
Comment (Tobychev) [reply docs/development/ceps/cep-001.rst]: changed. 

I think some repetition between the abstract and the introduction is unavoidable, but I now use the statement you like more in both places.

**Comment by kosack** [comment]
For the documentation, I suggest we  separate CEPs by *accepted*/*rejected*,  at least in the index of all CEPs, maybe even better in the CEP folder structure itself (refile them as necessary).   Reading some of the others like PIGs, I found it confusing to know if each was accepted or not.  

Also, the version where the CEP was *implemented* (if applicable) would also be useful, so perhaps a table like

| CEP  | Title | Status | Implemented in|
|----|----|----|-----|
| 00X | the title here | ACCEPTED | N/A | 
| 00Y | another title here | ACCEPTED | v0.19.1 |
| 00Z | a bad idea| REJECTED | | 
| 00N | the title here | ACCEPTED |  |

**Comment by Tobychev** [comment]
> For the documentation, I suggest we separate CEPs by _accepted_/_rejected_, at least in the index of all CEPs, maybe even better in the CEP folder structure itself (refile them as necessary). Reading some of the others like PIGs, I found it confusing to know if each was accepted or not without opening each one and reading.

I think the numpy solution of putting them in sections after status is enough, and likewise just updating the title of a CEP to include the version something was implemented is probably sufficient. 

I pushed a change to where things got moved to subdirectories implementing this.

**Comment by maxnoe** [comment]
Why do we need a directory for drafts? I think we only merge a CEP once it is proposed? Or even only when it is actually accepted or rejected?

So we would only need `accepted` and `rejected` directories?

File: docs/development/ceps/proposed/cep-001.rst [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+`PEP 481`_ is an example of a "process" PEP, proposing to move CPython development 
+to git and Github. 
+
+Submitting a CEP
+================
+Once a complete draft for a CEP is written, it can be submitted for comunity discussion 
+by opening a pull request adding a RST file to the  ``docs/development/ceps`` folder 
+in the main ctapipe repository.
+
+Each CEP must have a advocate -- someone who beyond writing the CEP using the style and
```
Comment (kosack) [inline docs/development/ceps/proposed/cep-001.rst:154]: _grammar_: Each CEP must have an advocate

**Review by kosack — APPROVED** [review]
Looks good.  It's true that the Drafts folder may be superfluous, but it depends a bit on the review process (if its fully in PRs only or decoupled)

**Comment by Tobychev** [comment]
@maxnoe Well the draft directory is simply where I put the draft as I work on them. It just seems sensible to keep drafts close to where they will end up once they are far along that they can be proposed, so no real harm in having it there to encourage others to keep things neat. 

If we really get problems with people committing lots of drafts by mistake then we could just add it to a .gitignore.

**Comment by maxnoe** [comment]
Fine with me in theory, but in practice it creates problems now....

docs build failes due to:

```
/home/runner/work/ctapipe/ctapipe/docs/development/ceps/index.rst:16: WARNING: toctree glob pattern 'accepted/*' didn't match any documents
/home/runner/work/ctapipe/ctapipe/docs/development/ceps/index.rst:33: WARNING: toctree glob pattern 'drafts/*' didn't match any documents
```

Is there a sphinx option to not error for empty globs?

**Review by kosack — DISMISSED** [review]

**Comment by Tobychev** [comment]
I just comment out the globbing for the two failing directory, not very elegant but it sort of works for now?

**Comment by maxnoe** [comment]
> I just comment out the globbing for the two failing directory, not very elegant but it sort of works for now?

I think there is nothing really we can do about it, so I would be fine.


@kosack Since we both reviewed and approved here, shall we directly add this to the "accepted" ceps?

**Review by kosack — APPROVED** [review]

**Comment by maxnoe** [comment]
Build is failing now due to the move, need to adapt which globs are commented out:

```
/home/runner/work/ctapipe/ctapipe/docs/development/ceps/index.rst:25: WARNING: toctree glob pattern 'proposed/*' didn't match any documents
looking for now-outdated files... none found
pickling environment... done
/home/runner/work/ctapipe/ctapipe/docs/development/ceps/accepted/cep-001.rst: WARNING: document isn't included in any toctree
```

really annoying...

**Review by kosack — APPROVED** [review]

**Review by maxnoe — APPROVED** [review]

**Pull request merged**